File: Source/Urho3D/CSharp/Managed/Math/EaseMath.cs
// Copyright (c) 2015 Xamarin Inc.
// This work is licensed under the terms of the MIT license.
// For a copy, see <https://opensource.org/licenses/MIT> or the accompanying LICENSE file.

using System;

namespace Urho3DNet
{
    public static class EaseMath
    {
        public static float BackIn(float time)
        {
            const float overshoot = 1.70158f;

            return time * time * ((overshoot + 1) * time - overshoot);
        }

        public static float BackOut(float time)
        {
            const float overshoot = 1.70158f;

            time = time - 1;
            return time * time * ((overshoot + 1) * time + overshoot) + 1;
        }

        public static float BackInOut(float time)
        {
            const float overshoot = 1.70158f * 1.525f;

            time = time * 2;
            if (time < 1)
            {
                return (time * time * ((overshoot + 1) * time - overshoot)) / 2;
            }
            else
            {
                time = time - 2;
                return (time * time * ((overshoot + 1) * time + overshoot)) / 2 + 1;
            }
        }

        public static float BounceOut(float time)
        {
            if (time < 1 / 2.75)
            {
                return 7.5625f * time * time;
            }
            else if (time < 2 / 2.75)
            {
                time -= 1.5f / 2.75f;
                return 7.5625f * time * time + 0.75f;
            }
            else if (time < 2.5 / 2.75)
            {
                time -= 2.25f / 2.75f;
                return 7.5625f * time * time + 0.9375f;
            }

            time -= 2.625f / 2.75f;
            return 7.5625f * time * time + 0.984375f;
        }

        public static float BounceIn(float time)
        {
            return 1f - BounceOut(1f - time);
        }

        public static float BounceInOut(float time)
        {
            if (time < 0.5f)
            {
                time = time * 2;
                return (1 - BounceOut(1 - time)) * 0.5f;
            }
            return BounceOut(time * 2 - 1) * 0.5f + 0.5f;
        }

        public static float SineOut(float time)
        {
            return (float)Math.Sin(time * MathDefs.PiOver2);
        }

        public static float SineIn(float time)
        {
            return -1f * (float)Math.Cos(time * MathDefs.PiOver2) + 1f;
        }

        public static float SineInOut(float time)
        {
            return -0.5f * ((float)Math.Cos((float)Math.PI * time) - 1f);
        }

        public static float ExponentialOut(float time)
        {
            return time == 1f ? 1f : (-(float)Math.Pow(2f, -10f * time / 1f) + 1f);
        }

        public static float ExponentialIn(float time)
        {
            return time == 0f ? 0f : (float)Math.Pow(2f, 10f * (time / 1f - 1f)) - 1f * 0.001f;
        }

        public static float ExponentialInOut(float time)
        {
            time /= 0.5f;
            if (time < 1)
            {
                return 0.5f * (float)Math.Pow(2f, 10f * (time - 1f));
            }
            else
            {
                return 0.5f * (-(float)Math.Pow(2f, -10f * (time - 1f)) + 2f);
            }
        }

        public static float ElasticIn(float time, float period)
        {
            if (time == 0 || time == 1)
            {
                return time;
            }
            else
            {
                float s = period / 4;
                time = time - 1;
                return -(float)(Math.Pow(2, 10 * time) * Math.Sin((time - s) * MathDefs.Pi * 2.0f / period));
            }
        }

        public static float ElasticOut(float time, float period)
        {
            if (time == 0 || time == 1)
            {
                return time;
            }
            else
            {
                float s = period / 4;
                return (float)(Math.Pow(2, -10 * time) * Math.Sin((time - s) * MathDefs.Pi * 2f / period) + 1);
            }
        }

        public static float ElasticInOut(float time, float period)
        {
            if (time == 0 || time == 1)
            {
                return time;
            }
            else
            {
                time = time * 2;
                if (period == 0)
                {
                    period = 0.3f * 1.5f;
                }

                float s = period / 4;

                time = time - 1;
                if (time < 0)
                {
                    return (float)(-0.5f * Math.Pow(2, 10 * time) * Math.Sin((time - s) * MathDefs.TwoPi / period));
                }
                else
                {
                    return (float)(Math.Pow(2, -10 * time) * Math.Sin((time - s) * MathDefs.TwoPi / period) * 0.5f + 1);
                }
            }
        }

    }
}

File: Source/Tests/Math/EaseMath.cpp
//
// Copyright (c) 2022-2022 the rbfx project.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR rhs
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR rhsWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR rhs DEALINGS IN
// THE SOFTWARE.
//

#include "../CommonUtils.h"
#include <Urho3D/Math/EaseMath.h>

using namespace Urho3D;

TEST_CASE("BackIn")
{
    CHECK(Equals(-0.0f, BackIn(0.0f)));
    CHECK(Equals(-0.064136565f, BackIn(0.25f)));
    CHECK(Equals(-0.087697506f, BackIn(0.5f)));
    CHECK(Equals(0.1825903f, BackIn(0.75f)));
    CHECK(Equals(1.0f, BackIn(1.0f)));
}

TEST_CASE("BackOut")
{
    CHECK(Equals(0.0f, BackOut(0.0f)));
    CHECK(Equals(0.8174097f, BackOut(0.25f)));
    CHECK(Equals(1.0876975f, BackOut(0.5f)));
    CHECK(Equals(1.0641365f, BackOut(0.75f)));
    CHECK(Equals(1.0f, BackOut(1.0f)));
}

TEST_CASE("BackInOut")
{
    CHECK(Equals(-0.0f, BackInOut(0.0f)));
    CHECK(Equals(-0.09968184f, BackInOut(0.25f)));
    CHECK(Equals(0.5f, BackInOut(0.5f)));
    CHECK(Equals(1.0996819f, BackInOut(0.75f)));
    CHECK(Equals(1.0f, BackInOut(1.0f)));
}

TEST_CASE("BounceOut")
{
    CHECK(Equals(0.0f, BounceOut(0.0f)));
    CHECK(Equals(0.47265625f, BounceOut(0.25f)));
    CHECK(Equals(0.765625f, BounceOut(0.5f)));
    CHECK(Equals(0.97265625f, BounceOut(0.75f)));
    CHECK(Equals(1.0f, BounceOut(1.0f)));
}

TEST_CASE("BounceIn")
{
    CHECK(Equals(0.0f, BounceIn(0.0f)));
    CHECK(Equals(0.02734375f, BounceIn(0.25f)));
    CHECK(Equals(0.234375f, BounceIn(0.5f)));
    CHECK(Equals(0.52734375f, BounceIn(0.75f)));
    CHECK(Equals(1.0f, BounceIn(1.0f)));
}

TEST_CASE("BounceInOut")
{
    CHECK(Equals(0.0f, BounceInOut(0.0f)));
    CHECK(Equals(0.1171875f, BounceInOut(0.25f)));
    CHECK(Equals(0.5f, BounceInOut(0.5f)));
    CHECK(Equals(0.8828125f, BounceInOut(0.75f)));
    CHECK(Equals(1.0f, BounceInOut(1.0f)));
}

TEST_CASE("SineOut")
{
    CHECK(Equals(0.0f, SineOut(0.0f)));
    CHECK(Equals(0.38268346f, SineOut(0.25f)));
    CHECK(Equals(0.70710677f, SineOut(0.5f)));
    CHECK(Equals(0.9238795f, SineOut(0.75f)));
    CHECK(Equals(1.0f, SineOut(1.0f)));
}

TEST_CASE("SineIn")
{
    CHECK(Equals(0.0f, SineIn(0.0f)));
    CHECK(Equals(0.076120496f, SineIn(0.25f)));
    CHECK(Equals(0.29289323f, SineIn(0.5f)));
    CHECK(Equals(0.6173166f, SineIn(0.75f)));
    CHECK(Equals(1.0f, SineIn(1.0f)));
}

TEST_CASE("SineInOut")
{
    CHECK(Equals(-0.0f, SineInOut(0.0f)));
    CHECK(Equals(0.14644662f, SineInOut(0.25f)));
    CHECK(Equals(0.5f, SineInOut(0.5f)));
    CHECK(Equals(0.8535534f, SineInOut(0.75f)));
    CHECK(Equals(1.0f, SineInOut(1.0f)));
}

TEST_CASE("ExponentialOut")
{
    CHECK(Equals(0.0f, ExponentialOut(0.0f)));
    CHECK(Equals(0.8232233f, ExponentialOut(0.25f)));
    CHECK(Equals(0.96875f, ExponentialOut(0.5f)));
    CHECK(Equals(0.9944757f, ExponentialOut(0.75f)));
    CHECK(Equals(1.0f, ExponentialOut(1.0f)));
}

TEST_CASE("ExponentialIn")
{
    CHECK(Equals(0.0f, ExponentialIn(0.0f)));
    CHECK(Equals(0.0045242715f, ExponentialIn(0.25f)));
    CHECK(Equals(0.03025f, ExponentialIn(0.5f)));
    CHECK(Equals(0.17577669f, ExponentialIn(0.75f)));
    CHECK(Equals(0.999f, ExponentialIn(1.0f)));
}

TEST_CASE("ExponentialInOut")
{
    CHECK(Equals(0.00048828125f, ExponentialInOut(0.0f)));
    CHECK(Equals(0.015625f, ExponentialInOut(0.25f)));
    CHECK(Equals(0.5f, ExponentialInOut(0.5f)));
    CHECK(Equals(0.984375f, ExponentialInOut(0.75f)));
    CHECK(Equals(0.9995117f, ExponentialInOut(1.0f)));
}

TEST_CASE("ElasticIn")
{
    CHECK(Equals(0.0f, ElasticIn(0.0f, 2.1f)));
    CHECK(Equals(-0.0034443287f, ElasticIn(0.25f, 2.1f)));
    CHECK(Equals(0.0023353111f, ElasticIn(0.5f, 2.1f)));
    CHECK(Equals(0.12958647f, ElasticIn(0.75f, 2.1f)));
    CHECK(Equals(1.0f, ElasticIn(1.0f, 2.1f)));
}

TEST_CASE("ElasticOut")
{
    CHECK(Equals(0.0f, ElasticOut(0.0f, 2.1f)));
    CHECK(Equals(0.87041354f, ElasticOut(0.25f, 2.1f)));
    CHECK(Equals(0.9976647f, ElasticOut(0.5f, 2.1f)));
    CHECK(Equals(1.0034443f, ElasticOut(0.75f, 2.1f)));
    CHECK(Equals(1.0f, ElasticOut(1.0f, 2.1f)));
}

TEST_CASE("ElasticInOut")
{
    CHECK(Equals(0.0f, ElasticInOut(0.0f, 2.1f)));
    CHECK(Equals(0.0011676556f, ElasticInOut(0.25f, 2.1f)));
    CHECK(Equals(0.5f, ElasticInOut(0.5f, 2.1f)));
    CHECK(Equals(0.99883235f, ElasticInOut(0.75f, 2.1f)));
    CHECK(Equals(1.0f, ElasticInOut(1.0f, 2.1f)));
}
